test/unit/slave/test_cluster_slave.py

```diff
@@ -237,6 +244,34 @@ def test_execute_subjob_passes_base_executor_index_to_executor(self):
 
         executor.execute_subjob.assert_called_with(1, 2, [], 12)
 
+    @genty_dataset(
+        responsive_master=(True,1,),
+        unresponsive_master=(False,1,),
+        unresponsive_master_retry=(False,2,),
+    )
+    def test_heartbeat_sends_post_to_master(self, is_master_responsive, heartbeat_failure_threshold):
+        expected_slave_heartbeat_url = 'http://{}/v1/slave/1/heartbeat'.format(self._FAKE_MASTER_URL)
+        Configuration['heartbeat_failure_threshold'] = heartbeat_failure_threshold
+
+        slave = self._create_cluster_slave()
+        slave.connect_to_master(self._FAKE_MASTER_URL)
+        slave.kill = Mock()
+        if not is_master_responsive:
+            self.mock_network.post_with_digest.side_effect = requests.ConnectionError
+
+        slave._run_heartbeat()
```

Comment:
Same comment as above about testing private methods. If you can't find a way around that, maybe at least have a supporting test that verifies `start_heartbeat_thread` sets up `slave._run_heartbeat` to run periodically. (Currently it looks like you have no tests that hit the public methods you added to the service classes.)

I didn't try writing this so it may be harder than I think, but my first instinct would be to 
- patch out SafeThread so that it executes its target synchronously
- set configuration so that the Nth post call will fail
- set configuration (or stub calls to datetime.now) so that there is no delay between scheduler calls
- assert that calling `start_heartbeat_thread` ends up sending N-1 post calls to the master

Anyway, just a suggestion. This is more of a test robustness issue than an actual issue with the feature code.